### PR TITLE
Add noisynet

### PR DESCRIPTION
**Pull request by MrSyee**
result graph : https://app.wandb.ai/khsyee/reinforcement_learning_examples/reports?view=khsyee%2Fnoisynet_iqn_test
result video : 
    LunarLander: https://youtu.be/LhsO8wHkhjY

Squashed commit of the following:

commit f13389ccc5717a49fadaf7417af027e4e9a43936
Author: khkim <kh.kim@medipixel.io>
Date:   Fri Apr 5 18:38:17 2019 +0900

    Unuse non-thread safe method (torch.rand)

commit 52258de93ae58f3abcfdc0fea039ad66b74e01d1
Author: khkim <kh.kim@medipixel.io>
Date:   Fri Apr 5 16:38:39 2019 +0900

    Add reset noise before calculating double q loss

commit f5dfad28df3b7c338bb6e8498097f97d8e33d1ca
Author: khkim <kh.kim@medipixel.io>
Date:   Fri Apr 5 09:56:28 2019 +0900

    Add comment

commit 03a818149074e28f3b4da6ed266a7dcdd7009774
Author: khkim <kh.kim@medipixel.io>
Date:   Thu Apr 4 12:21:31 2019 +0900

    Modify code for NoisyNet

commit ff2b23b99b637003d968274b4fd8537cd50a81e6
Author: khkim <kh.kim@medipixel.io>
Date:   Wed Apr 3 19:36:05 2019 +0900

    Add noisynet on fd

commit 792d0641b1574d1c02fa9431b26e34d687252175
Author: khkim <kh.kim@medipixel.io>
Date:   Wed Apr 3 17:21:15 2019 +0900

    Add reset noise and modify REAME.md

commit f982e328fcfa2f1ed36abb0536481f38c14b3f7a
Author: khkim <kh.kim@medipixel.io>
Date:   Wed Apr 3 14:40:21 2019 +0900

    Add noisynet on dqn, eval method on abstract agent

### CODE REVIEW

File: algorithms/dqn/networks.py [hunk 1]
```diff
@@ -19,6 +19,72 @@
...
+        out_features (int): output size of linear module
+        std_init (float): initial std value
+        weight_mu (nn.Parameter): mean value weight parameter
+        weight_sigma (nn.Parameter): std value weight parameter
+        bias_mu (nn.Parameter): mean value bias parameter
+        bias_sigma (nn.Parameter): std value bias parameter
+
+    """
+
+    def __init__(self, in_features: int, out_features: int, std_init: float = 0.5):
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: Add documentation for methods

**Comment by Curt-Park** [comment]
Why don't you set `use_noisy_net` flag when calculating dqn loss?

**Comment by medipixel** [comment]
This pull request **introduces 3 alerts** when merging fd26162ef04e5ee24070b12bcd8530c00ecd771c into a42151c5ecf7fb415d18d5121d2518209e29f479 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-52eab0f2244b85d567b0db44bc4c8e900571332e)

**new alerts:**

* 3 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by medipixel** [comment]
This pull request **introduces 3 alerts** when merging 45e033cb6b3968f0434a64a65e9539fb889d898b into a42151c5ecf7fb415d18d5121d2518209e29f479 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-20dbca3d06794880ebab0236acc35d5eed123b7e)

**new alerts:**

* 3 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by MrSyee** [comment]
result video : https://youtu.be/LhsO8wHkhjY

File: algorithms/common/networks/mlp.py [hunk 2]
```diff
@@ -31,6 +31,14 @@ def concat(
     return in_concat
 
 
+def init_layer_uniform(layer: nn.Module, init_w: float = 3e-3) -> nn.Module:
```
Comment (Curt-Park) [inline algorithms/common/networks/mlp.py]: nn.Module -> nn.Linear

File: algorithms/dqn/networks.py [hunk 3]
```diff
@@ -14,11 +14,95 @@
...
+
+        self.bias_mu = nn.Parameter(torch.Tensor(out_features))
+        self.bias_sigma = nn.Parameter(torch.Tensor(out_features))
+        self.register_buffer("bias_epsilon", torch.Tensor(out_features))
+
+        self.reset_parameters()
+        self.reset_noise()
+
+    def reset_parameters(self):
+        """Reset trainable parameters of networks (factorized gaussian noise)."""
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: parameters of networks -> network parameters

File: algorithms/dqn/networks.py [hunk 4]
```diff
@@ -14,11 +14,95 @@
...
+            return F.linear(
+                x,
+                self.weight_mu + self.weight_sigma * self.weight_epsilon,
+                self.bias_mu + self.bias_sigma * self.bias_epsilon,
+            )
+        else:
+            return F.linear(x, self.weight_mu, self.bias_mu)
+
+
+class NoisyWrapper:
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: Documentation
type annotation

File: algorithms/dqn/networks.py [hunk 5]
```diff
@@ -14,11 +14,95 @@
...
+    """Wrapper for changing hyper parameters of NoisyLinear"""
+
+    def __init__(self, noisy_linear, std_init=0.5):
+        """Initialization."""
+        self.NoisyLinear = noisy_linear
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: remove this line

File: algorithms/dqn/networks.py [hunk 6]
```diff
@@ -14,11 +14,95 @@
...
+        self.std_init = std_init
+
+    def __call__(self, *args: int):
+        """Return NoisyLinear instance set hyper parameters"""
+        return self.NoisyLinear(*args, self.std_init)
```
Comment (Curt-Park) [inline algorithms/dqn/networks.py]: return NoisyLinear(*args, std_init)

File: algorithms/dqn/networks.py (hunk 4 above)
Comment (Curt-Park) [reply algorithms/dqn/networks.py]: How about NoisyLinearConstructor

**Comment by medipixel** [comment]
This pull request **introduces 3 alerts** when merging 72bdcf93ee0cd10b9e4c5f711ce29204cddfa406 into a42151c5ecf7fb415d18d5121d2518209e29f479 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-70067cb46921e36120ebe24ea46a1884f4b3e858)

**new alerts:**

* 3 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: algorithms/dqn/utils.py [hunk 7]
```diff
@@ -145,7 +145,11 @@ def calculate_c51_loss(
     delta_z = float(v_max - v_min) / (atom_size - 1)
 
     with torch.no_grad():
+        # According to noisynet paper,
+        # Re-sample noisy net parameter on online network when using double q
```
Comment (Curt-Park) [inline algorithms/dqn/utils.py]: it resamples noisynet parameters ...

File: algorithms/dqn/utils.py [hunk 8]
```diff
@@ -145,7 +145,11 @@ def calculate_c51_loss(
...
+        # But, don't implement because performance issue
```
Comment (Curt-Park) [inline algorithms/dqn/utils.py]: , but we don't resample because there is no remarkable difference in performance.

File: algorithms/dqn/utils.py [hunk 9]
```diff
@@ -194,7 +198,11 @@ def calculate_dqn_loss(
     states, actions, rewards, next_states, dones = experiences[:5]
 
     q_values = model(states)
+    # According to noisynet paper,
+    # Re-sample noisy net parameter on online network when using double q
```
Comment (Curt-Park) [inline algorithms/dqn/utils.py]: same

**Comment by medipixel** [comment]
This pull request **introduces 3 alerts** when merging c0d68a2f4b5de2c16e4ef1c736b691a69dd75ccc into a42151c5ecf7fb415d18d5121d2518209e29f479 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-70ed494c21b678b50f51e5464487ec9d51475902)

**new alerts:**

* 3 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: algorithms/dqn/agent.py [hunk 10]
```diff
@@ -113,7 +115,11 @@ def select_action(self, state: np.ndarray) -> np.ndarray:
 
         # epsilon greedy policy
         # pylint: disable=comparison-with-callable
-        if not self.args.test and self.epsilon > np.random.random():
+        if (
+            not self.args.test
+            and self.epsilon > np.random.random()
+            and not self.hyper_params["USE_NOISY_NET"]
```
Comment (Curt-Park) [inline algorithms/dqn/agent.py]: How about just setting epsilon to 0?

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 7d0b9c82d372041303653cdd5095844176daa3bd into a42151c5ecf7fb415d18d5121d2518209e29f479 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-dfb7c5c530c7b802ed4f01b95acc0b3a9353bae7)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 04ea541919d3c958187e1115805e63aced78282f into a42151c5ecf7fb415d18d5121d2518209e29f479 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-ec74d363743f3f4cf7de55fa401b2fbf40011534)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 42949eca9bb9059812f1d640c8a3c43e9e0e0e73 into 86333b59e98981a7e0a3e94e323c62a22153bb32 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-85f491751fab93a23fd77e6f6d2448f319a5dd10)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 656257f6701ff22c6f07b052583b6e06e0b5a2fe into 86333b59e98981a7e0a3e94e323c62a22153bb32 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-f73a766dc8e55c7bdac92b3f305f3cb0db604293)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 688b032ffaa05434f06b8b0b3278a778474d8996 into 86333b59e98981a7e0a3e94e323c62a22153bb32 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-9342d4932fbab4dbed840629d3f9a5440f68054d)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 9bfecb1f30adcb9fe74a73c56a6abad1fa914f58 into 86333b59e98981a7e0a3e94e323c62a22153bb32 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-ac66bd6126451ed7201bea1bd1b1dc46578da722)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Curt-Park** [comment]
The test is running on...
https://app.wandb.ai/curt-park/dqn/reports?view=curt-park%2FPong%20%28DQN%20%2F%20C51%20%2F%20IQN%20%2F%20IQN%20-double%20q%29

**Comment by medipixel** [comment]
This pull request **fixes 1 alert** when merging 821874ef5252bdea45d361ffcef79f784602d41d into 86333b59e98981a7e0a3e94e323c62a22153bb32 - [view on LGTM.com](https://lgtm.com/projects/g/medipixel/rl_algorithms/rev/pr-478dea4795ecc036982a388936a76d68ac8afcca)

**fixed alerts:**

* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Curt-Park** [comment]
@all-contributor add @MrSyee for code

**Comment by Curt-Park** [comment]
@all-contributors add @MrSyee for code

**Comment by Curt-Park** [comment]
@all-contributors add @MrSyee for code
